src/databricks/labs/remorph/transpiler/lsp/lsp_engine.py

```diff
@@ -35,12 +68,175 @@ def parse(cls, data: dict[str, Any]) -> _LSPRemorphConfigV1:
+    def _register_lsp_features(self):
+        for name, options, func in _LSP_FEATURES:
+            decorator = self.protocol.fm.feature(name, options)
+            wrapper = self._wrapper_for_lsp_feature(func)
+            decorator(wrapper)
+
+    def _wrapper_for_lsp_feature(self, func):
+        def wrapper(params):
+            return func(self, params)
+
+        return wrapper
+
+
+class ChangeManager(abc.ABC):
+
+    @classmethod
+    def apply(cls, source_code: str, changes: Sequence[TextEdit]) -> str:
+        lines = source_code.split("\n")
+        for change in changes:
+            lines = cls._apply(lines, change)
+        return "\n".join(lines)
+
+    @classmethod
+    def _apply(cls, lines: list[str], change: TextEdit) -> list[str]:
+        new_lines = change.new_text.split("\n")
+        # special case where change covers the entire source code
+        if (
+            change.range.start.line <= 0
+            and change.range.start.character <= 0
+            and change.range.end.line >= len(lines) - 1
+            and change.range.end.character >= len(lines[-1])
+        ):
+            return new_lines
+        result: list[str] = []
+        # keep lines before
+        if change.range.start.line > 0:
+            result.extend(lines[0 : change.range.start.line])
+        # special case where change covers full lines
+        if change.range.start.character <= 0 and change.range.end.character >= len(lines[change.range.end.line]):
+            pass
+        # special case where change is within 1 line
+        elif change.range.start.line == change.range.end.line:
+            old_line = lines[change.range.start.line]
+            if change.range.start.character > 0:
+                new_lines[0] = old_line[0 : change.range.start.character] + new_lines[0]
+            if change.range.end.character < len(old_line):
+                new_lines[-1] += old_line[change.range.end.character :]
+        else:
+            if change.range.start.character > 0:
+                old_line = lines[change.range.start.line]
+                new_lines[0] = old_line[0 : change.range.start.character] + new_lines[0]
+            if change.range.end.character < len(lines[change.range.end.line]):
+                old_line = lines[change.range.end.line]
+                new_lines[-1] += old_line[change.range.end.character :]
+        result.extend(new_lines)
+        # keep lines after
+        if change.range.end.line < len(lines) - 1:
+            result.extend(lines[change.range.end.line + 1 :])
+        return result
+
+
 class LSPEngine(TranspileEngine):
 
     @classmethod
     def from_config_path(cls, config_path: Path) -> LSPEngine:
         config, custom = cls._load_config(config_path)
-        return LSPEngine(config, custom)
+        return LSPEngine(config_path.parent, config, custom)
 
     @classmethod
     def _load_config(cls, config_path: Path) -> tuple[_LSPRemorphConfigV1, dict[str, Any]]:
```

Comment:
TODO: Once this is stable, we create it as part of application context  and use Blueprint `Installation.load_local` to parse the config file.